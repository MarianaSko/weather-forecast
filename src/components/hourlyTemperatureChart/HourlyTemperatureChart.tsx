import { Box, Typography } from '@mui/material';
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  Tooltip,
  ResponsiveContainer,
  CartesianGrid,
} from 'recharts';
import type { ForecastAPIResponse } from '../../store';

interface HourlyTemperatureChartProps {
  forecast: ForecastAPIResponse;
}

const HourlyTemperatureChart = ({ forecast }: HourlyTemperatureChartProps) => {
  const today = new Date().toLocaleDateString('en-CA');
  console.log(today);

  const data = forecast.list
    .filter((item) => item.dt_txt.startsWith(today))
    .map((item): { time: string; temp: number } => ({
      time: `${new Date(item.dt_txt).getHours()}:00`,
      temp: Math.round(item.main.temp),
    }));

  return (
    <Box>
      <Typography variant="h6" gutterBottom color="#1d272c">
        HOURLY FORECAST (TODAY)
      </Typography>
      <ResponsiveContainer width="100%" height={300}>
        <LineChart data={data}>
          <CartesianGrid stroke="#1e3f67" />
          <XAxis dataKey="time" />
          <YAxis unit="Â°C" />
          <Tooltip />
          <Line type="monotone" dataKey="temp" stroke="#2196f3" strokeWidth={2} dot={{ r: 4 }} />
        </LineChart>
      </ResponsiveContainer>
    </Box>
  );
};

export default HourlyTemperatureChart;
