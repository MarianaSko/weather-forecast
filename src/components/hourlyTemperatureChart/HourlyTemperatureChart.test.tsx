import { render, screen } from '@testing-library/react';
import HourlyTemperatureChart from './HourlyTemperatureChart';
import type { ForecastAPIResponse } from '../../store';
import '@testing-library/jest-dom';

const mockForecast: ForecastAPIResponse = {
  list: [
    {
      dt: 1715594400,
      dt_txt: new Date().toISOString().split('T')[0] + ' 12:00:00',
      main: { temp: 21 },
      weather: [{ description: 'Clear', icon: '01d' }],
    },
    {
      dt: 1715598000,
      dt_txt: new Date().toISOString().split('T')[0] + ' 15:00:00',
      main: { temp: 23 },
      weather: [{ description: 'Sunny', icon: '01d' }],
    },
  ],
};

beforeAll(() => {
  global.ResizeObserver = class {
    observe() {}
    unobserve() {}
    disconnect() {}
  };
});

describe('HourlyTemperatureChart', () => {
  it('renders chart title', () => {
    render(<HourlyTemperatureChart forecast={mockForecast} />);
    expect(screen.getByText(/HOURLY FORECAST \(TODAY\)/i)).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<HourlyTemperatureChart forecast={mockForecast} />);
    expect(container).toMatchSnapshot();
  });

  it('renders without crashing even if city is missing', () => {
    const noCityForecast = { ...mockForecast, city: undefined };
    render(<HourlyTemperatureChart forecast={noCityForecast} />);
    expect(screen.getByText(/HOURLY FORECAST/i)).toBeInTheDocument();
  });

  it('handles empty forecast list', () => {
    const emptyForecast: ForecastAPIResponse = {
      city: { name: 'Kyiv', country: 'UA' },
      list: [],
    };
    render(<HourlyTemperatureChart forecast={emptyForecast} />);
    expect(screen.getByText(/HOURLY FORECAST/i)).toBeInTheDocument();
  });
});
