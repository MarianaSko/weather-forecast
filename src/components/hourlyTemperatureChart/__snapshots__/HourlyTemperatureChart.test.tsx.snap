// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HourlyTemperatureChart matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-14ewf1c-MuiTypography-root"
    >
      HOURLY FORECAST (TODAY)
    </h6>
    <div
      class="recharts-responsive-container"
      style="width: 100%; height: 300px; min-width: 0;"
    />
  </div>
</div>
`;
